astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,851 @@
+        # zero. In this case, set a tolerance for the final velocity; if it's
+        # below this tolerance, assume the delta velocity is essentially nan.
+
+        fin_vel_mag = np.linalg.norm(fin_vel)
+
+        if NUMPY_LT_1_17:
+            fin_vel_mag *= fin_vel.unit
+
+        if np.isnan(delta_vel) or fin_vel_mag < 1e-7 * fin_vel.unit:
+            delta_vel = -self.radial_velocity
+
+        return self._apply_relativistic_doppler_shift(delta_vel)
+
+    def with_observer_velocity(self, frame, preserve_observer_frame=False):
+        """
+        Alters the velocity of the observer, but not the position.
+
+        If a coordinate frame is specified, the observer velocities will be
+        modified to be stationary in the specified frame. If a coordinate
+        instance is specified, optionally with non-zero velocities, the
+        observer velocities will be updated so that the observer is co-moving
+        with the specified coordinates.
+
+        Parameters
+        ----------
+        frame : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The observation frame containing the new velocity for the observer.
+        preserve_observer_frame : bool
+            If `True`, the final observer frame class will be the same as the
+            original one, and if `False` it will be the frame of the velocity
+            reference class.
+
+        Returns
+        -------
+        new_coord : `SpectralCoord`
+            The new coordinate object representing the spectral data
+            transformed based on the observer's new velocity frame.
+        """
+
+        if self.observer is None or self.target is None:
+            raise ValueError("This method can only be used if both observer "
+                             "and target are defined on the SpectralCoord.")
+
+        if hasattr(frame, 'frame'):
+            frame = frame.frame
+        elif isinstance(frame, type):
+            frame_cls = frame
+            frame = frame_cls(0 * u.m, 0 * u.m, 0 * u.m,
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+        elif isinstance(frame, str):
+            frame_cls = frame_transform_graph.lookup_name(frame)
+            frame = frame_cls(0 * u.m, 0 * u.m, 0 * u.m,
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+
+        observer = update_differentials_to_match(self.observer, frame,
+                                                 preserve_observer_frame=preserve_observer_frame)
+
+        new_coord = self._change_observer_to(observer)
+
+        return new_coord
+
+    def with_los_shift(self, target_shift=None, observer_shift=None):
```

Comment:
Elsewhere, `line of sight` is always `radial`. Although very verbose, perhaps this should be 
```suggestion
    def with_radial_velocity_shift(self, target_shift=None, observer_shift=None):
```